### ISSUE TITLE
Structured Quantity Distribution

### ISSUE THREAD

**Issue by nstarman** [astropy/astropy#12909]
### Description

A structured Quantity cannot be put into a Distribution.

### Expected behavior

The Distribution constructor should work with a structured Quantity, like it does with a structured ndarray.

### Actual behavior

The reshaping of the samples array in the Distribution constructor runs afoul of the StructuredUnit dtype versus units shape and throws a ValueError.

### Steps to Reproduce

```python
from astropy.uncertainty import Distribution
import numpy as np
import astropy.units as u

x = np.array([(0, 0, 0.6), (0.1, -0.01, 0.59)], dtype=[("nu1", float), ("nu2", float), ("nu3", np.float128)])
q = x * u.Unit((u.eV, u.eV, u.eV))
Distribution(q)
```

### System Details

macOS-12.1-x86_64-i386-64bit
Python 3.9.10 | packaged by conda-forge | (main, Feb  1 2022, 21:28:27) 
[Clang 11.1.0 ]
Numpy 1.21.4
pyerfa 2.0.0.1
astropy 5.1.dev531+g23eb23183.d20220216
Scipy 1.7.1
Matplotlib 3.5.1

**Comment by nstarman** [comment]
ping @mhvk, do you think this is an easy fix? Or should this be a "known issue"?

**Comment by mhvk** [comment]
Not super-easy, I think. First step would be to think through carefully whether one has a distribution of structured elements, or each element is a distribution. I think that for an array that normally has 'x', 'y', and 'z', one would like to continue to be able to do `mc_array['x']`. That in turn suggests (but does not really demand), that each element should be a distribution, i.e., each part of the `dtype` becomes structured itself (with 'samples').

Best may be first to think through carefully an actual use case (like cosmology distributions, I suppose?!).

**Comment by mhvk** [comment]
p.s. A possible advantage of each element of a structure having 'samples' is that it may give the freedom to have samples only for some elements. On the other hand, the whole idea that one can do `.samples` easily then is destroyed, so perhaps the approach in #12908 is better.

**Comment by nstarman** [comment]
ping @mhvk. I think some of your recent PRs might address this?

**Comment by mhvk** [comment]
You're right, this issue was partially what made me push on this. Fixed by #15304.
